File: app/src/main/java/com/elementary/tasks/core/controller/EventControlFactory.java
package com.elementary.tasks.core.controller;

import com.elementary.tasks.core.data.models.Reminder;
import com.elementary.tasks.core.utils.LogUtil;

import androidx.annotation.NonNull;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public final class EventControlFactory {

    private static final String TAG = "EventControlFactory";

    private EventControlFactory() {
    }

    @NonNull
    public static EventControl getController(@NonNull Reminder reminder) {
        EventControl control;
        if (Reminder.isSame(reminder.getType(), Reminder.BY_DATE_SHOP)) {
            control = new ShoppingEvent(reminder);
        } else if (Reminder.isBase(reminder.getType(), Reminder.BY_DATE)) {
            control = new DateEvent(reminder);
        } else if (Reminder.isBase(reminder.getType(), Reminder.BY_LOCATION)) {
            control = new LocationEvent(reminder);
        } else if (Reminder.isBase(reminder.getType(), Reminder.BY_MONTH)) {
            control = new MonthlyEvent(reminder);
        } else if (Reminder.isBase(reminder.getType(), Reminder.BY_WEEK)) {
            control = new WeeklyEvent(reminder);
        } else if (Reminder.isBase(reminder.getType(), Reminder.BY_OUT)) {
            control = new LocationEvent(reminder);
        } else if (Reminder.isBase(reminder.getType(), Reminder.BY_PLACES)) {
            control = new LocationEvent(reminder);
        } else if (Reminder.isSame(reminder.getType(), Reminder.BY_TIME)) {
            control = new TimerEvent(reminder);
        } else if (Reminder.isBase(reminder.getType(), Reminder.BY_DAY_OF_YEAR)) {
            control = new YearlyEvent(reminder);
        } else {
            control = new DateEvent(reminder);
        }
        LogUtil.d(TAG, "getController: " + control);
        return control;
    }
}

File: app/src/main/java/com/elementary/tasks/core/controller/EventControlFactory.kt
package com.elementary.tasks.core.controller

import android.content.Context
import com.github.naz013.appwidgets.AppWidgetUpdater
import com.elementary.tasks.core.data.ui.reminder.UiReminderType
import com.elementary.tasks.core.services.JobScheduler
import com.elementary.tasks.core.utils.GoogleCalendarUtils
import com.elementary.tasks.core.utils.Notifier
import com.github.naz013.common.datetime.DateTimeManager
import com.elementary.tasks.core.utils.datetime.RecurEventManager
import com.elementary.tasks.core.utils.params.Prefs
import com.github.naz013.domain.Reminder
import com.github.naz013.common.TextProvider
import com.github.naz013.logging.Logger
import com.github.naz013.repository.GoogleTaskRepository
import com.github.naz013.repository.ReminderRepository
import com.github.naz013.ui.common.datetime.ModelDateTimeFormatter

class EventControlFactory(
  private val prefs: Prefs,
  private val googleCalendarUtils: GoogleCalendarUtils,
  private val context: Context,
  private val notifier: Notifier,
  private val jobScheduler: JobScheduler,
  private val appWidgetUpdater: AppWidgetUpdater,
  private val textProvider: TextProvider,
  private val reminderRepository: ReminderRepository,
  private val googleTaskRepository: GoogleTaskRepository,
  private val dateTimeManager: DateTimeManager,
  private val recurEventManager: RecurEventManager,
  private val modelDateTimeFormatter: ModelDateTimeFormatter
) {

  fun getController(reminder: Reminder): EventControl {
    val type = UiReminderType(reminder.type)
    return when {
      type.isSame(Reminder.BY_DATE_SHOP) && !reminder.hasReminder -> {
        ShoppingEvent(
          reminder,
          reminderRepository,
          prefs,
          googleCalendarUtils,
          notifier,
          jobScheduler,
          appWidgetUpdater,
          textProvider,
          dateTimeManager,
          googleTaskRepository
        )
      }

      type.isBase(UiReminderType.Base.DATE) -> {
        DateEvent(
          reminder,
          reminderRepository,
          prefs,
          googleCalendarUtils,
          notifier,
          jobScheduler,
          appWidgetUpdater,
          textProvider,
          dateTimeManager,
          googleTaskRepository
        )
      }

      type.isGpsType() -> {
        LocationEvent(
          reminder,
          reminderRepository,
          prefs,
          context,
          notifier,
          jobScheduler,
          appWidgetUpdater,
          dateTimeManager
        )
      }

      type.isBase(UiReminderType.Base.MONTHLY) -> {
        MonthlyEvent(
          reminder,
          reminderRepository,
          prefs,
          googleCalendarUtils,
          notifier,
          jobScheduler,
          appWidgetUpdater,
          textProvider,
          dateTimeManager,
          googleTaskRepository,
          modelDateTimeFormatter
        )
      }

      type.isBase(UiReminderType.Base.WEEKDAY) -> {
        WeeklyEvent(
          reminder,
          reminderRepository,
          prefs,
          googleCalendarUtils,
          notifier,
          jobScheduler,
          appWidgetUpdater,
          textProvider,
          dateTimeManager,
          googleTaskRepository,
          modelDateTimeFormatter
        )
      }

      type.isBase(UiReminderType.Base.TIMER) -> {
        TimerEvent(
          reminder,
          reminderRepository,
          prefs,
          googleCalendarUtils,
          notifier,
          jobScheduler,
          appWidgetUpdater,
          textProvider,
          dateTimeManager,
          googleTaskRepository,
          modelDateTimeFormatter
        )
      }

      type.isBase(UiReminderType.Base.YEARLY) ->
        YearlyEvent(
          reminder,
          reminderRepository,
          prefs,
          googleCalendarUtils,
          notifier,
          jobScheduler,
          appWidgetUpdater,
          textProvider,
          dateTimeManager,
          googleTaskRepository,
          modelDateTimeFormatter
        )

      type.isBase(UiReminderType.Base.RECUR) ->
        RecurEvent(
          reminder,
          reminderRepository,
          prefs,
          googleCalendarUtils,
          notifier,
          jobScheduler,
          appWidgetUpdater,
          textProvider,
          dateTimeManager,
          googleTaskRepository,
          recurEventManager
        )

      else -> DateEvent(
        reminder,
        reminderRepository,
        prefs,
        googleCalendarUtils,
        notifier,
        jobScheduler,
        appWidgetUpdater,
        textProvider,
        dateTimeManager,
        googleTaskRepository
      )
    }.also {
      Logger.d("getController: $it")
    }
  }
}
